### PR TITLE
Windiator creationclass And HotFixes

### PR DESCRIPTION
**Pull request by Veltouille**
# Windiator creationclass And HotFixes

## New Features

**New Champion : Windiator**
- This new champion is built with the respect of the architecture.
- The PR contains all the needed scripts to load and play this champion.

**AutoAttackWindiator**
- The auto attack of the Windiator.
- Able to launch the MeleAttack associted to the weapon.
- Can be contacted by the MeleAttack on the weapon if something is hit.
- When it hit a monster, has a chance to activate the PassiveWindiator.

**PassiveWindiator**
- Passive Spell of the Windiator.
- Contains a method that is called buy the AutoAttack when then random is a success.
- This method applies a WindSpiritFavorStatus to an ally in a certain range.

**WindStrikeSpell**
- Spell1 of the Windiator.
- Instanciate a WindStrikeSwing prefab with a WindStrikeSwing script associated
- Contacted by the WindStrikeSwing to apply damages and a WindPushStatus.

**TridentLaunchSpell**
- Spell2 of the Windiator.
- Launch a Trident prefab with a Trident script associated.
- Contacted by the Trident script when a Monster is hit to apply damages.

**StormWallSpell**
- Spell3 of the Windiator.
- Whe launched, instanciate a StormWall prefab with a StormWallBehaviour associated.
- When contacted by the StormWallBehaviour, applies a BackWindStatus (on Enemy Entities) or an InsideStormStatus (on Ally Entities)). 

**GustSpell**
- Spell4 of the Windiator.
- When launched, get all players in a certain range and applies a GustStatus to each one.

**WindSpiritFavorStatus**
- Applied by the PassiveWindiator.
- Reduces cooldowns of the entity.
- **For the moment, use as a Debug.Log because we don't have the properties on players.**

**WindPushStatus**
- Applied by the WindStrikeSpell.
- Knock up enemies in the air at a certain distance.
- While enemies are in the air they can not attack or move.
- **For the moment, use as a Debug.Log because we don't have the properties on Monsters.**

**BackWindStatus**
- Status Applied by the StormWallSpell.
- When applied, the entity (Projectile or LivingBase) is slowed down during an amount of time.
- **For the moment, use as a Debug.Log (on players only) because we don't have the movement speed  property on Monsters.**

**InsideStormStatus**
- Status Applied by the StormWallSpell.
- When applied, the entity (Projectile or LivingBase) is accelerated during an amount of time.
- **For the moment, use as a Debug.Log (on players only) because we don't have the movement speed  property on players.**

**GustStatus**
- Status applied by the GustSpell
- Applied to all allies at the same time.
- Reduces cooldowns of the entity.
- Increases Movement Speed and Damages of the entity.
- **For the moment, use as a Debug.Log because we don't have the properties on players.**

**MeleAttack**
- Mother Abstract Class of all meleAttacks.
- Able to swap a collider(trigger) on the weapon (enable/disable).
- Able to detect if a something is in the trigger.

**IWeaponAttack**
- Interface that regroups all MeleAttacks.

**WindiatorSimpleAttack**
- Is a MeleAttack.
- Associated to the weapon of the Windiator.
- Is able to play animations and detect if we hit an Enemy.
- Able to contact the AutoAttackWindiator if an enemy is hit.

**WindStrikeSwing**
- Script associated to the WindStrikeSwing prefab instanciated by the WindStrikeSpell.
- Able to be resized from the JSON Spell.
- Lauch a spin attack in front of the player.
- For each entity caught in the spin, we contact the WindStrikeSpell.

**StormWallBehaviour**
- Script associated to the StormWall prefab instanciated by the StormWallSpell.
- Moves forward at a certain speed and slow down to the last 33% of the distance.
- When an ally or enemy (LivingBase or Projectile) passes through the wall, it contacts the StormWallSpell.

**Trident**
- Script associated to the Trident prefab instanciated by the TridentLaunchSpell.
- Linear projectile that contacts the TridentLaunchSpell when it hit something.

## Enhancements

**PassivesBase**
- Now passives are able to load and apply Status
- Now have a method LoadResource that is used to dynamically load prefabs that will be instanciated

**LinearProjectiles**
- Now get the tag AllyEntity or EnemyEntity when instanciated by an Ally or an Enemy.
- Now contains a method to be accelarated by x%.
- Now contains a method to be slowed down by x%.

**AutoAttackBase**
- Now able to apply Status.
- Now have a method LoadResource that is used to dynamically load prefabs that will be instanciated

**Spell**
- Now have a method LoadResource that is used to dynamically load prefabs that will be instanciated

**Champion and ChampionData**
- Was previously called Character and CharacterData.
- All classes associated where renamed (classe, inner class and JSON file)
- Now have a property that displays the name of the Champion.

## BugFixes
- Delete useless gameObject in the Canvas > GUI > Spell.
- Windiator is now correctly able to load its spells only one time.
- Now champions are correctly loaded inside all scene. This is due to modifications of the loading order of a champion.
- Monsters and NPC have now a correct way to attach a rigidobody to themselves with the good properties.
- The animation of the Windiator's auto-attack is not looping anymore.
- StatusBase now correctly handles reset of Status and is able to detect if the Status is on a player (GUI modified) or not.

## MinorFixes
- Deleting some Debug.Log
- Fixing some variables name to fit with the C# rules.

## Lot of Tasks

[Task n° 147](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=147&project_id=3)
[Task n° 148](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=148&project_id=3)
[Task n° 174](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=174&project_id=3)
[Task n° 176](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=176&project_id=3)
[Task n° 179](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=179&project_id=3)
[Task n° 188](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=188&project_id=3)
[Task n° 192](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=192&project_id=3)
[Task n° 195](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=195&project_id=3)
[Task n° 193](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=193&project_id=3)
[Task n° 195](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=195&project_id=3)
[Task n° 198](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=198&project_id=3)
[Task n° 200](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=200&project_id=3)
[Task n° 202](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=202&project_id=3)
[Task n° 204](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=204&project_id=3)
[Task n° 205](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=205&project_id=3)
[Task n° 206](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=206&project_id=3)
[Task n° 207](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=207&project_id=3)
[Task n° 210](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=210&project_id=3)
[Task n° 211](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=211&project_id=3)
[Task n° 216](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=216&project_id=3)

## Activity Diagram

[Windiator](http://www.plantuml.com/plantuml/svg/hPPTQzim58Rl_IiEb49oCzAlCMmOJTEkBQGiD1szrkmJHwij6UaufOpzznt9SUmu7_tOSf4OwSXwtqVl4Jct90oLUHOK_2j9QWMZJ5WxqYh6Wk1t43vAbKX1slY9A3WOotHErmPHXHCUaGjS7vGpw0rBqaCY4J_rWUQee7ThVrXC-W6g9C3CegjohvwIaK-uhshAU-UF8ils5JqOcQ2YiIXLFDzL2AxokhIqiw1wOiEpu4yGU3g1BJ99BK1BqyncAmg0y1e911YX4fq32t-1cT4vpmWOIqkWPsuK9fbuGGDIGIfol_BWHjD4ggTmM1IyolCff97qJIoqcPAWqW9B6F8x8w_LgNJwH2gaQebi3g2Zivbm6yfPxqOIJCiuHcjdvOencHBxNC1kAYs9o4MAVlSxhIoXYQhfqLnmLe0q54R7K3GSAevSCdKSOQHrbkXd1I4SVJe6Yt5jWRyho3k8JmlCigEEaIOybH7eSMZHqMJOoaCkF60lap_Eg7L6sUPxV5f-f-LqGSplhqntpcbAfMeEP_So7x7UJF1FjUSUGeCseErdxB4STx2ikYKAmdkCIQWqGn06nHgQAdirEPSdqCRdQcBqBrudUTLtnNpJ5CsLOaERoBR4TLBQUON29s3TS6tNRNcbCFTx-eUFRthh8NlNh-mRZWNttfUAwnRYnsyaVj8XldRrr4G_6hQcRpOLVdwZmDEEG7VZyHMzpiqDDfbjhYFc5rvenSFtc9JSs2q1fuF3e9gxqmlCsRJR41EUEbjDNNfFSA4rqsQiPuVHQ-XF1k-v3TPZvhpMpZO3-3ntaUfrbZp87E4h3FeWBVJA8X5KdykbjE9dvfkMZCw0xt38QxFSDwmeHLgUad_xtlEuu7kskgYwMZipBR6TcJrYRrNCVSrHNs2cOqal_o9rev7QQhTssLBlviAzVF-Zv5Q6MrgRqPR8Lkc7W4QTWYhbDkgugaEznS8grkVyn7-v_W80)

### CODE REVIEW

**Comment by Myrilandel** [comment]
Belle PR avec pleins de tasks :D Je commence la review d'ici peu.
J'aimerai qu'on fixe le bug du double chargement de spells avant de merge car c'est un problème qu'on ne rencontrait pas avant ([#211](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=211&project_id=3))
Aussi, j'aimerai qu'on ajoute la [#202](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=202&project_id=3) pour être sur de pas avoir de problèmes par la suite ;) (Je m'occupe des deux et je push sur cette branche)

Edit : Comme vu avec toi ce matin, est-ce que tu pourras rajouter le lien vers le diagramme d'activité stp ? :)

File: Assets/Scripts/Champions/BASE_CHAMPION/Spell.cs [hunk 1]
```diff
@@ -28,9 +28,10 @@ public abstract class Spell : MonoBehaviour, IDisplayable
     public string[] OtherValues { get; protected set; }
     public GameObject[] Status { get; protected set; }
     public string[] Description { get; protected set; }
+    public int NumberOfStacks { get; protected set; }
 
     public bool HasGCD;
-    public int NumberOfStacks;
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/BASE_CHAMPION/Spell.cs:33]: Je me demande si on devrait pas rapatrier les StackableSpells directement dans les spells.
Faire une fusion des deux classes et avoir une détection d'un bool IsStackable depuis un JSON.
Ca éviterai de devoir décider de mettre a chaque fois un extends Spell ou StackableSpell (on limite les erreurs et on se prend pas le choux)
Qu'est-ce que tu en penses ?

File: Assets/Scripts/Champions/MeleAttack.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        _arme = GetComponent<Collider>();
+        _arme.enabled = false;
+    }
+
+    /// <summary>
+    /// OnTriggerEnter Unity Method
+    /// Will be called in the child class to Check if the Weapon touch an enemy
+    /// </summary>
+    /// <param name="other">is an Enemy (EntityLivingBase)</param>
+    protected void OnTriggerEnter(Collider other)
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/MeleAttack.cs:31]: Je ne comprends pas pourquoi cette méthode n'est ni virtual ni abstract ? 
Est-ce que c'est volontaire ?

File: Assets/Scripts/Champions/MeleAttack.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    public void SwapEnableArmeCol()
+    {
+        _arme.enabled = !_arme.enabled;
+    }
+
+    /// <summary>
+    /// Getter of the Collider's weapon
+    /// </summary>
+    /// <returns>if the weapon Collider is enable or not</returns>
+    public bool GetValueColArme()
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/MeleAttack.cs:53]: Est-ce que tu penses qu'on pourrait remplacer cette méthode Get par une propriété get{}; set{} ?
Je sais qu'elles sont moins gourmandes

File: Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+/// <summary>
+/// AutoAttackWindiator extends AutoAttackBase
+/// This is obviously the script that launch the AutoAttacks
+/// This script is attached to the Windiator Character
+/// </summary>
+public class AutoAttackWindiator : AutoAttackBase {
+
+    private Animator _anim;
+    private WindiatorSimpleAttack _wsa;
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs]: Je chipote mais je suis pas trop fan des noms de variable acronymes :D . A l'occase est-ce qu'on pourra mettre un 
_windiatorSimpleAttack ?

File: Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        if (AutoAttackIsReady())
+        {
+            base.AutoAttack();
+
+            if (!_wsa.GetValueColArme())
+            {
+                _wsa.SwapEnableArmeCol();
+            }
+
+            _anim.SetTrigger("AutoAttack");
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs]: Bien ouej ;)

File: Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/GustStatus.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/GustStatus.cs:4]: Bien qu'elle soit vide, est-ce qu'on pourrait mettre des commentaires dans cette classe stp ? 
Ainsi qu'un debug.Log pour qu'on sache à terme ce que devrai faire ce Status. Ca nous évitera d'aller fouiner dans toutes les archi pour retrouver son but

File: Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        origin = transform.position;
+        transform.parent = null;
+        transform.position += new Vector3(0, transform.lossyScale.y / 2, 0);
+
+        if (_parentSpell != null)
+        {
+            _maxRange = float.Parse(_parentSpell.OtherValues[0]);
+        }
+
+        Invoke("DestroyItself", float.Parse(_parentSpell.OtherValues[1]));
```
Comment (Myrilandel) [inline Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs]: L'invoke est une très bonne idée, on avait vu ensemble je me souviens.
Je m'étais renseigné par la suite et voilà ce que j'ai pu tirer de la doc. 
Les invokes sont très coûteuses mais sont très intéressantes parce qu'on peut les canceler. Par exemple Je pose un status avec un Invoke d'une methode qui me détruira ce Status au bout de x secondes. Si on reset le status alors on a just a CancelInvoke et reInvoke dans x secondes.

Dans notre cas, je pense qu'on pourrait simplement mettre un Destroy(GameObject obj, float time). A moins que tu comptes ajouter une mécanique qui puisse annuler la destruction du mur ? 
Qu'en penses tu ?

File: Assets/Scripts/Entities/EntityLivingBase.cs [hunk 8]
```diff
@@ -16,6 +16,46 @@ public abstract class EntityLivingBase : MonoBehaviour
...
+    /** Awake, protected virtual void
+	 * First of all, we are checking that all LivingEntities that are not players have a Rigidbody that fits with the game rules. Which are.
+	 * Living Entities must have a rigidbody which uses gravity and is not kinematic.
+	 * Position are not frozen.
+	 * Only X and Z Rotation are frozen.
+     * Be aware that RigidbodyConstraints are sum of bytes, this is why we always use a simple | in test.
+	 **/
+    protected virtual void Awake()
+    {
+        if (gameObject.tag != "Player")
```
Comment (Myrilandel) [inline Assets/Scripts/Entities/EntityLivingBase.cs:28]: J'ai pas prêté attention et je me suis trompé sur cette partie, j'ai mis tous les Debug.Log en français :D 
Je corrigerai quand on passera sur Sentinelle

**Review by Myrilandel — CHANGES_REQUESTED** [review]
Un très très gros sprint qui s'achève avec un énorme travail.
Beaucoup de refacto et du nouveaux éléments, la PR parle d'elle même. Tu as fais un travail monstre sur le champion et tous les bugFixes associés. On a vraiment quelque chose de clair, et d'optimisé pour la suite donc gros GG. 
J'ai laissé quelques commentaires tu verras qu'il n'y a que très peu de choses à changer (la plupart du temps des commentaires/nom de variables).
Je passe en Request Changes le temps d'avoir mes réponses. Encore gg pour le boulot accompli ;)

File: Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs (hunk 5 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs]: lol merci !

File: Assets/Scripts/Entities/EntityLivingBase.cs (hunk 8 above)
Comment (Veltouille) [reply Assets/Scripts/Entities/EntityLivingBase.cs:28]: Ok pas de soucis !

File: Assets/Scripts/Champions/BASE_CHAMPION/Spell.cs (hunk 1 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/BASE_CHAMPION/Spell.cs:33]: Je pense que ça peut faire une classe plus chargée mais que ça peut être pratique pour la suite =)

File: Assets/Scripts/Champions/MeleAttack.cs (hunk 2 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/MeleAttack.cs:31]: Je ne sais pas si tu te souviens mais j'avais un soucis en lançant le test qui me disait que je ne pouvait déclarer ma classe comme abstract. Donc, je l'ai corrigé pour que cela fonctionne.

A tester de nouveau mais avec l’apparition d'un véritable model, je ne sais pas comment ça va s'agencer.

Pour répondre à ta question, pour le moment c'est volontaire.

File: Assets/Scripts/Champions/MeleAttack.cs (hunk 3 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/MeleAttack.cs:53]: Je ne sais pas, pour moi c'est facile à tester si tu veux faire un :

         if (!_wsa.GetValueColArme())
         {
             _wsa.SwapEnableArmeCol();
         }

Mais si c'est possible avec un "get set" je n'ai juste pas eu l'idée donc pourquoi pas =)

File: Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs (hunk 4 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/Windiator/AutoAttack/AutoAttackWindiator.cs]: Selon votre bon vouloir, Messire Buluto !

File: Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs (hunk 7 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs]: > Dans notre cas, je pense qu'on pourrait simplement mettre un Destroy(GameObject obj, float time). A moins que tu comptes ajouter une mécanique qui puisse annuler la destruction du mur ?
Qu'en penses tu ?

Maintenant que tu en parles pourquoi pas ajouter certaines méca, mais pas maintenant. Donc soit on a un PC méga puissant soit on change avec la solution du Destroy(GameObject obj, float time) ?

**Review by Veltouille — COMMENTED** [review]
J'ai répondu, je vais certaines modif ce soir !

File: Assets/Scripts/Entities/EntityLivingBase.cs [hunk 9]
```diff
@@ -98,11 +138,11 @@ public void InstantKill()
...
-			s.DestroyStatus();
-		}
+        IStatus[] status = GetComponentsInChildren<IStatus>();
+        foreach (IStatus s in status)
+        {
+            s.DestroyStatus();
+        }
         /// TODO : We need to add HERE the death Animation.
         StartCoroutine(DespawnEntity());
     }
```
Comment (Veltouille) [inline Assets/Scripts/Entities/EntityLivingBase.cs:148]: Super les tests de correction d'entity et de son rigidbody =)
A mettre en Anglais pour enfoncer le clou ^^

D'ailleurs et sauf erreur de ma part j'ai pas vu de "Rigidbite" MDR.

**Review by Veltouille — COMMENTED** [review]
Ta partie !

File: Assets/Scripts/Champions/MeleAttack.cs (hunk 2 above)
Comment (Myrilandel) [reply Assets/Scripts/Champions/MeleAttack.cs:31]: Okay ca marche en fait ce que je trouve bizarre c'est que tu puisses la déclarer sans Virtual et que tu la redefinisse dans les classes filles sans messages d'avertissement. On regardera plus tard c'est pas gênant pour le moment !

File: Assets/Scripts/Champions/MeleAttack.cs (hunk 3 above)
Comment (Myrilandel) [reply Assets/Scripts/Champions/MeleAttack.cs:53]: Le test sera identique normalement

File: Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs (hunk 7 above)
Comment (Myrilandel) [reply Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs]: L'idée c'est de se dire quel est le meilleur outil? 
Un Invoke permet de lancer une méthode dans X secondes avec possibilité de cancel ce lancement, c'est gourmand.
Un Destroy au bout de x seconde ne peut pas se cancel mais c'est moins gourmand. Les deux sont viables mais comme je te disais l'autre fois, si on économise 1% de CPU ici, on peut utiliser 1% de plus sur une mécanique plus interessante :)

File: Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/GustStatus.cs (hunk 6 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/GustStatus.cs:4]: Vu que c'est toi qui à fait le Gust mdr !

File: Assets/Scripts/Entities/EntityLivingBase.cs (hunk 9 above)
Comment (Myrilandel) [reply Assets/Scripts/Entities/EntityLivingBase.cs:148]: Je n'ai pas osé :D

File: Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs (hunk 7 above)
Comment (Veltouille) [reply Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs]: Donc je peux changer la méthode pour le moment, on verra plus tard !

Comment (Myrilandel) [reply Assets/Scripts/Champions/Windiator/ScriptsOnPrefabs/StormWallBehaviour.cs]: Ca roule !

**Comment by Veltouille** [comment]
Voilà, j'ai fait les derniers changes : 

J'ai modifié la variable `_wsa;`
j'ai passé une variable `public Bool colArme { get; set; }`
J'ai changé l'Invoke par un Destroy.

Merci pour ta review, merci pour avoir bosser avec moi sur mes tâches, c'était vraiment plus intéressant,
et merci pour ton message pour la PR ça fait vraiment plaisir !

On va cartonner la suite du PJV !
Tu peux merger quand tu veux.

**Review by Myrilandel — APPROVED** [review]

**Pull request merged**